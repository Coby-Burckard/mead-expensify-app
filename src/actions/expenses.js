import uuid from "uuid";
import database from "../firebase/firebase"

// ADD_EXPENSE
const addExpense = (expense) => ({
  type: "ADD_EXPENSE",
  expense
});

const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0
    } = expenseData

    const expense={
      description,
      note,
      amount,
      createdAt
    }

    const uid = getState().auth.uid

    return database.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }))
    })
  }
}

// REMOVE_EXPENSE
const removeExpense = ({ id } = {}) => {
  const action = {
    type: "REMOVE_EXPENSE",
    id
  }
  return action
};

const startRemoveExpense = ( { id } = {}) => {
  return (dispatch, getState) => { 
    const uid = getState().auth.uid 
    database.ref(`users/${uid}/expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({id}))
    }).catch((error) => {
      console.log('Error deleting expense: ', error)
    })
  }
}

// EDIT_EXPENSE
const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates
});

const startEditExpense = (id, expenseData) => {
  const {
    description = "",
    note = "",
    amount = 0,
    createdAt = 0
  } = expenseData

  const updates = {description, note, amount, createdAt}

  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`expenses/${id}`)
      .set(updates)
      .then(() => {
        dispatch(editExpense(id, updates))
      })
      .catch((error) => {
        console.log('Error updating db: ', error)
      })
  }
}

// SET_EXPENSES
const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

// START_SET_EXPENSES
const startSetExpenses = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses`)
      .once('value')
      .then((snapshot) => {
        const expenses = []
        snapshot.forEach(expense => {
          expenses.push({
            id: expense.key,
            ...expense.val()
          })
        })
        dispatch(setExpenses(expenses))
      })
      .catch('failed initial read from DB')
  }
}

export { addExpense, removeExpense, editExpense, setExpenses, startAddExpense, startSetExpenses, startRemoveExpense, startEditExpense };
