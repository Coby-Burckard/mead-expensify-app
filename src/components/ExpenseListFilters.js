import React from "react";
import { connect } from "react-redux";
import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate
} from "../actions/filters";
import { DateRangePicker } from "react-dates";

class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.dispatch(setStartDate(startDate));
    this.props.dispatch(setEndDate(endDate));
  };

  onFocusChange = calendarFocused => {
    this.setState(() => ({ calendarFocused }));
  };

  render() {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input
              className="text-input"
              placeholder="search..."
              type="text"
              value={this.props.filters.text}
              onChange={e => {
                this.props.dispatch(setTextFilter(e.target.value));
              }}
            />
          </div>
          <div className="input-group__item">
            <select
              className="select"
              value={this.props.filters.sortBy}
              onChange={e => {
                this.props.dispatch(
                  e.target.value === "date" ? sortByDate() : sortByAmount()
                );
              }}
            >
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker
              startDate={this.props.filters.startDate}
              startDateId="hi there"
              endDate={this.props.filters.endDate}
              endDateId="end here"
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.calendarFocused}
              onFocusChange={this.onFocusChange}
              numberOfMonths={1}
              isOutsideRange={() => false}
              showClearDates={true}
            />
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => {
  return {
    filters: state.filters
  };
};

export default connect(mapStateToProps)(ExpenseListFilters);
