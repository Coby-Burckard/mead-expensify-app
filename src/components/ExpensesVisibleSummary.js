import { connect } from 'react-redux'
import numeral from 'numeral'
import React from 'react'
import getVisibleExpenses from '../selectors/expenses'
import getExpensesTotal from '../selectors/selectedExpensesTotal'

export const ExpensesVisibleSummary = (props) => {
    const expenseWord = props.expensesCount === 1 ? ' expense ' : ' expenses '
    const formattedTotal = numeral(props.expensesTotal/100).format('$0,0.00')
    return (
        <div>
            <span>
                Viewing {props.expensesCount} {expenseWord} totalling {formattedTotal}
            </span>
        </div>
    )
}

const mapStateToProps = (state => {
    const visibleExpenses = getVisibleExpenses(state.expenses, state.filters)
    return {
        expensesCount: visibleExpenses.length,
        expensesTotal: getExpensesTotal(visibleExpenses)
    }
})

export default connect(mapStateToProps)(ExpensesVisibleSummary)