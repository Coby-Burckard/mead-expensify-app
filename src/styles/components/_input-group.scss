.input-group{
    display: flex;
    margin-bottom: $m-size;
    flex-direction: column;

    @media (min-width: $desktop-breakpoint) {
        flex-direction: row;
        margin-bottom: $l-size;
    }
}

.input-group__item {
    margin: 0 0 $m-size 0;
    padding: auto $s-size;

    @media (min-width: $desktop-breakpoint) {
        margin: 0 $s-size 0 0 
    }
}

