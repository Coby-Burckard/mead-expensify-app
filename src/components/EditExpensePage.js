import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { editExpense, removeExpense } from "../actions/expenses";

const EditExpensePage = props => {
  return (
    <div>
      <ExpenseForm
        onSubmit={expense => {
          props.dispatch(editExpense(props.match.params.id, expense));
          props.history.push("/");
        }}
        expense={props.expense}
      />
      <button
        onClick={() => {
          props.dispatch(removeExpense({ id: props.match.params.id }));
          props.history.push("/");
        }}
      >
        Remove
      </button>
    </div>
  );
};

const mapStatetoProps = (state, props) => {
  return {
    expense: state.expenses.find(
      expense => expense.id === props.match.params.id
    )
  };
};

export default connect(mapStatetoProps)(EditExpensePage);
