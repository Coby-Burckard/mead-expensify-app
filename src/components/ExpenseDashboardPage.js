import React from "react";
import ExpensesList from "./ExpensesList";
import ExpenseListFilters from "./ExpenseListFilters";
import ExpensesVisibleSummary from "./ExpensesVisibleSummary"

const ExpenseDashboardPage = () => (
  <div>
    <ExpenseListFilters />
    <ExpensesVisibleSummary />
    <ExpensesList />
  </div>
);

export default ExpenseDashboardPage;
