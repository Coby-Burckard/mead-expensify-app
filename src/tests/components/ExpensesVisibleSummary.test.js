import React from 'react'
import { shallow } from 'enzyme'
import { ExpensesVisibleSummary } from '../../components/ExpensesVisibleSummary'
import expenses from '../fixtures/expenses'

test('should render statement for 0 expenses', ()=> {
    const expensesCount = 0
    const expensesTotal = 0
    const wrapper = shallow(<ExpensesVisibleSummary expensesCount={expensesCount} expensesTotal={expensesTotal}/>)
    expect(wrapper).toMatchSnapshot()
})

test('should render statement for 1 expense', ()=> {
    const expensesCount = 1
    const expensesTotal = 12678.55
    const wrapper = shallow(<ExpensesVisibleSummary expensesCount={expensesCount} expensesTotal={expensesTotal}/>)
    expect(wrapper).toMatchSnapshot()
})

test('should render statement for multiple expenses', ()=> {
    const expensesCount = 4
    const expensesTotal = 102.55
    const wrapper = shallow(<ExpensesVisibleSummary expensesCount={expensesCount} expensesTotal={expensesTotal}/>)
    expect(wrapper).toMatchSnapshot()
})
