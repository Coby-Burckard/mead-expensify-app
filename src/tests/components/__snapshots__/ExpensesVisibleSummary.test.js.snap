// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render statement for 0 expenses 1`] = `
<div>
  <span>
    Viewing 
    0
     
     expenses 
     totalling 
    $0.00
  </span>
</div>
`;

exports[`should render statement for 1 expense 1`] = `
<div>
  <span>
    Viewing 
    1
     
     expense 
     totalling 
    $126.79
  </span>
</div>
`;

exports[`should render statement for multiple expenses 1`] = `
<div>
  <span>
    Viewing 
    4
     
     expenses 
     totalling 
    $1.03
  </span>
</div>
`;
